### PR TITLE
Port Ordinal Ignore Case Optimization changes

### PR DESCRIPTION
**Pull request by tarekgh**
This is porting the changes from the PR https://github.com/dotnet/runtime/pull/40910.

### CODE REVIEW

**Comment by tarekgh** [comment]
https://github.com/dotnet/runtime/issues/37919
https://github.com/dotnet/runtime/issues/26054
https://github.com/dotnet/runtime/issues/40942

**Comment by tarekgh** [comment]
@safern @GrabYourPitchforks this is straight porting the changes you have approved to RC1 branch. 

CC @danmosemsft @ericstj @Anipik

**Review by Anipik — APPROVED** [review]

**Comment by danmosemsft** [comment]
Port approved

**Review by GrabYourPitchforks — APPROVED** [review]
Thanks Tarek! 😸

**Review by safern — APPROVED** [review]

File: src/libraries/Native/Unix/System.Globalization.Native/pal_casing.c [hunk 1]
```diff
@@ -150,6 +150,24 @@ void GlobalizationNative_ChangeCaseTurkish(
...
+    for (int i = 0; i < 256; i++)
+    {
+        // Unfortunately, to ensure one-to-one simple mapping we have to call u_toupper on every character.
+        // Using string casing ICU APIs cannot give such results even when using NULL locale to force root behavior.
+        pTarget[i] = (UChar) u_toupper((UChar32)(pageNumber + i));
+    }
+
+    if (pageNumber == 0x0100)
+    {
+        // Disable Turkish I behavior on Ordinal operations
```
Comment (ericstj) [inline src/libraries/Native/Unix/System.Globalization.Native/pal_casing.c:165]: I know this is just a clean port, but special cases like this that aren't driven by culture data make me nervous.  How do we know this is the only special case and there won't be more in the future?

Comment (tarekgh) [reply src/libraries/Native/Unix/System.Globalization.Native/pal_casing.c:165]: I added these for compatibility reason not for any other reason. I am not really worried if we get any cases in the future as these will be a new introduced character that never existed before and it is ok for us to decide to handle it either way.

Comment (tarekgh) [reply src/libraries/Native/Unix/System.Globalization.Native/pal_casing.c:165]: I forgot to mention, disabling these characters I think not related to the culture that these characters belongs to but related to mapping the character to the ASCII range. so it is not related to the cultures.

Comment (ericstj) [reply src/libraries/Native/Unix/System.Globalization.Native/pal_casing.c:165]: I see, the comment could use some clarification.  My understanding is that this is treating upper/lowercase the same for both the `Turkish lowercase i` and the `Latin small letter long s`.  In other words: these characters should have no upper-case representation, even if `u_toupper` returns an uppercase value for them.  Is that correct?

Comment (tarekgh) [reply src/libraries/Native/Unix/System.Globalization.Native/pal_casing.c:165]: yes that is correct.

**Comment by tarekgh** [comment]
The failure in the CI is tracked by the issue https://github.com/dotnet/runtime/issues/39009

https://helix.dot.net/api/2019-06-17/jobs/61043e95-3b03-4c8c-bf90-1b1b5cf6fb28/workitems/System.DirectoryServices.Protocols.Tests/console%20

```
ushd .
/home/helixbot/work/A434096D/p/dotnet exec --runtimeconfig System.DirectoryServices.Protocols.Tests.runtimeconfig.json --depsfile System.DirectoryServices.Protocols.Tests.deps.json xunit.console.dll System.DirectoryServices.Protocols.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing 
popd
===========================================================================================================
~/work/A434096D/w/A872094B/e ~/work/A434096D/w/A872094B/e
  Discovering: System.DirectoryServices.Protocols.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.DirectoryServices.Protocols.Tests (found 343 of 369 test cases)
  Starting:    System.DirectoryServices.Protocols.Tests (parallel test collections = on, max threads = 2)
dotnet: ../nptl/pthread_mutex_lock.c:79: __pthread_mutex_lock: Assertion `mutex->__data.__owner == 0' failed.
./RunTests.sh: line 161: 95592 Aborted                 "$RUNTIME_PATH/dotnet" exec --runtimeconfig System.DirectoryServices.Protocols.Tests.runtimeconfig.json --depsfile System.DirectoryServices.Protocols.Tests.deps.json xunit.console.dll System.DirectoryServices.Protocols.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing $RSP_FILE
~/work/A434096D/w/A872094B/e
----- end Tue Aug 18 17:35:36 UTC 2020 ----- exit code 134 ----------------------------------------------------------
exit code 134 means SIGABRT Abort. Managed or native assert, or runtime check such as heap corruption, caused call to abort(). Core dumped.
```

**Review by ericstj — APPROVED** [review]

**Pull request merged**